### DISCUSSION TITLE
How can I write a customized controller/estimator in python?

### DISCUSSION THREAD

**Discussion by jsuh9** [bitcraze/discussions#881] [Q&A]
It seems that Kalman Filter and PID are used for estimation/control.
I am searching through crazyflie-lib-python repository to see how they are implemented (model, gain values, etc.) because I would like to modify/customize but I could not locate anywhere in the repository how/where they are written (for example, I was looking for something along the lines of 'kalman_filter.py' or 'pid_control.py'). Am I looking at the wrong repository? Please help.

**Comment by krichardsson** [comment]
The controller and estimators are implemented in the [Crazyflie firmware](https://github.com/bitcraze/crazyflie-firmware), not the python lib. The python lib runs in a computer and is mainly used to communicate with the Crazyflie.

You can read about the [stabilizer module in the firmware documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/)

You can find a brief documentation of how to add controllers and estimators from a Crayflie app in the [OOT build documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/oot/)

The on-board controllers and estimators are mainly implemented in the `controller` and `estimator` directories in the [`src/module` part of the tree](https://github.com/bitcraze/crazyflie-firmware/tree/master/src/modules)
